### DISCUSSION TITLE
`TypeVar` bound or constrained to `bool`, diverging behavior with mypy

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#6914] [Q&A] [answered]
Following https://discuss.python.org/t/41483 where I couldn't find a viable solution (it would require typing features not yet implemented), I'm trying to find an alternative solution. TL;DR, `Field` is generic over the `__set__` and `__get__` type, and both can be `None` if `null=True` is used when instantiating the class.

The alternative I'm exploring is adding an extra type variable `_NullT`, bound to `bool`. What I did at first was:

```python
from typing_extensions import TypeVar

_NullT = TypeVar("_NullT", bound=bool, default=Literal[False])

class Field(Generic[_ST, _GT, _NullT]):
    def __init__(self, null: _NullT = False):
        ...

class IntField(Field[int | float, int, _NullT]):
    pass

f = IntField(null=True)
```

In this example, `f` is inferred as `IntField[bool]`, which makes sense! The type of `True` is `bool`. So I've tried switching to this type variable instead, using constraints:

```python
_NullT = TypeVar("_NullT", Literal[False], Literal[True], default=Literal[False])
```
As I understand it, solving this type variable would result in either `Literal[False]` or `Literal[True]` (which can be considered as types?). However, pyright still infers `f` as being `IntField[bool]`.

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.13&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFADJIwCmIAhgDZkDi1KtSAxmWAG60MwdACYAoUJFiJUaAPrUAHjRQBnJGFWZsufABVE1AGp0Qo0QGIos2SuoxrUXowCu1KQmqjZAZV1QAvFD6HsYgABQARD66EWScGjD0TiBIdCgw-rogrgCUFlayaHYOIHbOICjunrLMfoHBRiaRNTFxfCap6ZnZ1HleAHLODAx1QQahzYPDrRRUtIwA2lmuALpklDT0DAsAYoy2a1DC1MB0Qxkb89t7DAd55rJTIwFjIU1RTzMARmDOKML%2BH5gJhHE5nBgXOZbXb7agrPqcBh0FQqKA7JDUBjCMKsdgpTgLaJkFrEz7wgBcoig1NBwAKqCo1jCtgYwDIKCGDHJVk%2BLxuthylJpwqgADpxWZhZYdmBOM41OgoBoCkV7LIoMBcFAUGAAO5CmkAAT4AiEYmFxzp1lVTJZwG56MxwkJvmJtXWUMWyzhh1QKhgaU41G5AFkwMcQXq8dyiHAclAALQAPistSgAB8oP0NMGxRLhcb%2BCBBCIqTTLSrirJmZj7WiMViXbo3c3ZptFvyfWQ-QGUEHQ%2BHMTxddHCMR48nU7pueLRWXqWZRIjkaiAJLpR1YsKb52ofCZ4AlmDd9Kkzm6CnzqAIFeL5coqBhiOUq90wLrmA7sIc4bdXJmUp%2BEYWR4A8MIn0xMIclFYA8iAA)

```python
from typing import Any, Literal, Generic, overload
from typing_extensions import TypeVar

# __set__ value type
_ST = TypeVar("_ST", contravariant=True)
# __get__ return type
_GT = TypeVar("_GT", covariant=True)

_NullT = TypeVar("_NullT", Literal[True], Literal[False], default=Literal[False])
#_NullT = TypeVar("_NullT", bound=bool, default=Literal[False])

class Field(Generic[_ST, _GT, _NullT]):
    def __init__(self, null: _NullT = False):
        ...

    # Focusing on __get__ for now:
    @overload
    def __get__(self: Field[_ST, _GT, Literal[True]], instance: Model, owner: Any) -> _GT | None: ...
    @overload
    def __get__(self: Field[_ST, _GT, Literal[False]], instance: Model, owner: Any) -> _GT: ...
    


class IntField(Field[int | float, int, _NullT]):
    pass


class Model:

    f = IntField(null=True)

reveal_type(Model().f)
```

Now something really weird is, with either the first or second `_NullT` type variable commented out, I always get `IntField[bool]`, but if you leave both uncommented, `f` is inferred as `IntField[Literal[False]]` (ideally it should be `IntField[Literal[True]]`, but it doesn't seem to be taking into account the passed value of `null`). I believe this is because pyright gets confused with the redefinition of `_NullT`.

---

With `mypy`, I get something almost working, although there are still some weird errors happening. `f` is correctly inferred as `IntField[Literal[True/False]]` (depending on the initialized `null` value), however it does not seem to pick the right overload.

[mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=aeb3a2b763b057afc8a64cc0a7367c0a)

**Comment by Viicos** [comment]
https://github.com/python/mypy/issues/10207#issuecomment-798812236 gives a nice solution, but it would be even better if inferred without any `__init__` overloads. (And I can't make it work with subclasses when other type variables are in play: [playground](https://pyright-play.net/?pythonVersion=3.13&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSEgMYA0UAMkjGQIYA2NAKgK4ItEeJEAakxA0AqiiRgUNMADcyLMEwAmAKDXsoAXijsBwkAAoAROxM0ARmE4oV262BYBKTQEYdegyNPs3J1zUKFiYAZ1CoAEEjYlJyCgBtdg43AF1nAC41KByoAAF5RWV1XKgVImAoAH0q1AYao1CiFmAMqIT6RhBWBIAxVibUlKGoECIYThAUKomeIjbO5hY%2BgaJUzwA6LecoAFoAPigAOWl57NLcrY2NUoKFECVVc5zyypq6mAamlrbIjoYlkkQJw1sMaGMJlMZtxeAsAd1luxgWsdgdjqcshdLlsbrlXtVapJPlVGs1gODxpNprNYVBHCxPP0WE1UYcTqRMVioN9gBsIVToXNPPyoTSiLicncik9SviahRWCwvmTfv8uj0kSCRn50ntDlw5glUDAaMbUpysVcJflCg9is8yhUCQqWEqSTzVYsEStmaC9GlWZgUDALRcrQ65VUXW7SS1AxIpCgjcGODCiMmTUGYKlzQ7SkhKjy%2BZTRWnQ1zRiWUFAjG4aAAmVwVkXVtwaYJhCIAISMf2SWfSoYQnY0TA8umiLcFvG0mqIrjGClYMwERjHC6IS6V8AQRDXbiMzh2UAAxHo0xnTcHUqP655ohutyvd2vG2pF0Rlzu90x64fj2exoaJY45QD2U5irOyKPp%2B26riBzhAA))

**Reply by erictraut** [reply, marked as answer]
Looks like a bug in pyright. You're doing something very unusual here — using literal types as value constraints. Pyright's constraint solver includes some heuristics about when to retain literals and when to widen the type to non-literals. I'm guessing that those heuristics are failing in this case. Please file a bug, and I'll investigate further.

**Reply by Viicos** [reply]
> using literal types as value constraints.

Indeed, this seems to be the only way to avoid defining overloads (and it also plays really nice with type var defaults). Thanks for the answer, I've opened an issue.
